wireframe

SecondSon1 opened the pull request:
Первая КТ. Пока отображается только wireframe, объект подгружается из .obj. Пока нет culling, clipping, lighting, а также камера зафиксирована на (0, 0, 0) и смотрит в направлении (0, 0, -1)

Commit:
wireframe, no culling, no clipping, no lighting.

DimaTrushin commented on .clang-format:
```diff
@@ -0,0 +1,10 @@
+AllowShortLoopsOnASingleLine: false
+DerivePointerAlignment: true
+KeepEmptyLinesAtTheStartOfBlocks: true
+SortIncludes: false
```
Добавь пустую строку в конце каждого файла. Дело в том, что гит на это ругается и пользователь твоего репозитория будет видеть сообщения об ошибках от гита. Не делай так.
Добавить в clang-format правило:
InsertNewlineAtEOF: true

DimaTrushin commented on CMakeLists.txt:
Я смотрю требования заливать в пустую ветку прошло мимо тебя. Я НЕ МОГУ блять физически комментировать в ревью тот код, что уже лежит в master ветке. Как я блять должен это делать по твоему?Исправь, иначе я не могу ничего сделать.

DimaTrushin commented on CMakeLists.txt:
```diff
@@ -12,7 +12,7 @@ add_subdirectory("external/glog")
 
```
К строчке, блять, 6 выше:
Добавь строчку
```cpp
set(CMAKE_CXX_STANDARD_REQUIRED True)
```
А то cmake разрешено откатиться на более низкий стандарт.

DimaTrushin replied:
К строчке, блять, 7 выше.
Не надо жестко хардкодить дебаг сборку в cmake файле. Устанавливай этот атрибут при сборке. Если ты собираешь в IDE, то это ее задача передать нужный параметр режима компиляции. Если ты собираешь через консоль, то у тебя должны быть два скрипта для сборки в релизе и дебаге.

DimaTrushin commented on src/main.cpp:
```diff
@@ -1,26 +1,81 @@
+#include <scene/scene.hpp>
+#include <scene/from_obj.hpp>
 
 constexpr size_t WINDOW_WIDTH = 1280;
```
Если вы определяете
- переменные
- функции
- классы
в cpp файле, то все это надо обязательно завернуть в anonymous namespace. Иначе это чревато ошибками линкера.

DimaTrushin replied:
НЕ используй стиль CAPS_ONLY ни для чего кроме макросов. Это очень сбивает при чтении кода. Если тебе нужно обратить внимание на то, что переменная -- это константное значение, добавь префикс k или k_ в начале.

DimaTrushin commented on src/main.cpp:
```diff
@@ -1,26 +1,81 @@
+#include <numbers>
+#include <chrono>
+#include <cmath>
 
```
Какого хуя у тебя хедеры разбиты в два блока? Если тебе важен порядок хедеров, то это признак плохого дизайна. Это надо делать НЕ так. Слей в один список и пусть clang-format их сам упорядочивает.

DimaTrushin commented on src/main.cpp:
```diff
@@ -1,26 +1,81 @@
+  return mesh.Transform(translationMatrix * rotationMatrixX * rotationMatrixZ);
+}
+
 int main(int argc, char *argv[]) {
```
У тебя не отключены exception, потому надо поставить блок на них. Это делается так
```cpp
#include "Application.h"
#include "Except.h"

int main() {
  try {
    renderer::Application app;
    app.run();
  } catch(...) {
    except::react();
  }
  return 0;
}
Except.h
namespace except {
void react();
}
Except.cpp
#include "Except.h"

namespace except {
void react() {
  try {
    throw;
  } catch(std::exception& e) {
    // обрабатываешь известные исключения
  } catch(...) {
    // обрабатываешь незивестные исключения
  }
}
}
```
Функция main -- это главный оркестровщик. Она всю работу делегирует другому коду (даже обработку исключений). Вторая функция main -- следить, чтобы исключения не убежали в OS. Это важно потому что ты не хочешь путать два вида ошибок исполнения: ошибки в программе из-за самой программы и ошибки в программе из-за операционной среды. Не надо засорять OS непойми какими сообщениями о своих фейлах.

DimaTrushin commented on src/main.cpp:
```diff
@@ -1,26 +1,81 @@
+  std::filesystem::current_path(PATH_TO_PROJECT_ROOT);
+  LOG(INFO) << "cwd: " << std::filesystem::current_path();
 
   SDL::Window window(WINDOW_TITLE, WINDOW_WIDTH, WINDOW_HEIGHT, 0);
```
Подобные вызовы библиотечного кода надо заворачивать внутрь объектов, которые предоставляют тебе нужный интерфейс, а не работать с сырыми системными вызовами, которые хуй пойми что значат.

DimaTrushin commented on src/main.cpp:
```diff
@@ -1,26 +1,81 @@
+}
+
 int main(int argc, char *argv[]) {
   google::InitGoogleLogging(argv[0]);
```
Если у тебя идет инициализация глобальных объектов, поставь вызов перед конструктором Application функции по инициализации всех глобальных объектов.

DimaTrushin commented on src/main.cpp:
```diff
@@ -1,26 +1,81 @@
   google::InitGoogleLogging(argv[0]);
+  FLAGS_stderrthreshold = google::GLOG_INFO;  // for debugging
+
+  std::filesystem::current_path(PATH_TO_PROJECT_ROOT);
```
Зачем этот код тут? Какая его задача?

DimaTrushin commented on src/main.cpp:
```diff
@@ -1,26 +1,81 @@
 
   SDL::Window window(WINDOW_TITLE, WINDOW_WIDTH, WINDOW_HEIGHT, 0);
-  SDL::Renderer renderer = window.GetRenderer(nullptr);
+  SDL::Renderer sdlRenderer = window.GetRenderer(nullptr);
```
Вот это все детали конструктора какого-то объекта по типу View и ты вытащил сюда наружу его кишки.

DimaTrushin commented on src/main.cpp:
```diff
@@ -1,26 +1,81 @@
-  SDL::Renderer renderer = window.GetRenderer(nullptr);
+  SDL::Renderer sdlRenderer = window.GetRenderer(nullptr);
+
+  auto color_pixel = [&sdlRenderer](size_t x, size_t y, uint8_t r, uint8_t g, uint8_t b) {
```
Какого хрена у тебя эта штука перехватывает sdlRenderer по ссылке? Такого кода быть не должно в финальной версии. Надо отличать код по построению картинки и код по отображению картинки. Это код по отображению картинки? Если да, то перенести его внутрь view объекта в метод отрисовки.

DimaTrushin commented on src/main.cpp:
```diff
@@ -1,26 +1,81 @@
 int main(int argc, char *argv[]) {
   google::InitGoogleLogging(argv[0]);
+  FLAGS_stderrthreshold = google::GLOG_INFO;  // for debugging
+
```
Выдели основной объект для поддержания контекста твоего приложения, скажем Application. Тогда у тебя должен быть основной объект
```cpp
class Application {
public:
  Application();
  void run();
private:
  Model model_;
  View view_;
  Controller controller_;
};
```
Задача класса Application -- инициализировать все ресурсы для работы и соединить все свои компоненты в конструкторе. И в методе run запустить event loop. Все.

Renderer -- это часть ядра Application, которое находится внутри Model. Опять же это если ты используешь MVC паттерн. Если ты не хочешь его использовать. Ты можешь поместить его прямо в Application. Тогда там будет
```cpp
class Application {
public:
  ...
private:
  Renderer renderer_;
  World world_;
  Camera camera_;
  Screen screen_;
  View view_;
};
```
Задача Application в этом случае передавать данные между компонентами. То есть в event loop, если происходит какое-то действие, на которое надо поменять сцену, то Application должен ее обновить и запустить пайплайн отрисовки как-то так
```cpp
// где-то внутри event loop
camera_.Move(...);
screen_ = renderer_.Render(world_, camera_);
view_.show(screen_);
```
Что-то в этом духе. Более того, в этом случае screen_ даже не нужно хранить в самом Application. Его можно сделать локальной переменной event loop. Если ты хочешь переиспользовать память в screen_, то можно сделать так
```cpp
screen_ = renderer_.Render(world_, camera_, std::move(screen_));
```
То есть ты муваешь внутрь метода отрисовки существующий экран, и там уже метод отрисовки смотрит, если его можно переиспользовать, то он переиспользует и возвращает данные в экран.

DimaTrushin commented on src/main.cpp:
```diff
@@ -1,26 +1,81 @@
-  SDL_RenderPresent(renderer.renderer);
+  auto start = std::chrono::high_resolution_clock::now();
 
   bool shouldClose = false;
```
Код со строки 80 до 88 -- это код для event loop-а твоего приложения.

DimaTrushin replied:
Обычно булевы флаги -- это признак ошибки дизайна. В целом флаги -- это ошибки дизайна.

DimaTrushin commented on src/main.cpp:
```diff
@@ -31,6 +86,18 @@ int main(int argc, char *argv[]) {
         break;
```
Комментарий к, блять, строчкам номер 85-86. Плохой дизайн event loop-а. Ты не завершаешь цикл по брейку, ты лишь вываливаешься в больший цикл, где сначала делаешь какую-то работу, и лишь потом вываливаешься. Я бы выходил из цикла по return. Но для этого этот цикл надо выделить в отдельную функцию. Заодно пропадет ебанутый булев флаг.

DimaTrushin commented on src/main.cpp:
```diff
@@ -31,6 +86,18 @@ int main(int argc, char *argv[]) {
+
+    SDL_SetRenderDrawColor(sdlRenderer.renderer, 0, 0, 0, 0);
+    SDL_RenderClear(sdlRenderer.renderer);
+    auto elapsed = std::chrono::high_resolution_clock::now() - start;
```
Сделай временные якоря и wrapper-ы для замерения времени.

DimaTrushin commented on src/main.cpp:
```diff
@@ -31,6 +86,18 @@ int main(int argc, char *argv[]) {
+    SDL_RenderClear(sdlRenderer.renderer);
+    auto elapsed = std::chrono::high_resolution_clock::now() - start;
+    auto elapsedInSeconds =
+        std::chrono::duration_cast<std::chrono::microseconds>(elapsed).count() / 1000000.0;
```
Не работай с сырыми std::chrono типами. Заведи для работы с временем отдельный объект Time или как-нибудь так его назови. И чтобы он возвращал тебе время в секундах или микросекундах по вызову функций вида
```cpp
auto anchor = Time::now();
...
double elapsed = anchor.elapst_microseconds();
```
Не делай низкоуровневую работу там, где ты хочешь думать большими мазками на высоком уровне.

DimaTrushin commented on src/main.cpp:
```diff
@@ -31,6 +86,18 @@ int main(int argc, char *argv[]) {
+    rendererObject.Render(sceneToRender, sdlRenderer);
+
+    SDL_RenderPresent(sdlRenderer.renderer);
+    SDL_Delay(16);
```
НИКОГДА ни при каких условиях НЕ делай delay в thread-е обработчике сообщений.

DimaTrushin commented on src/main.cpp:
```diff
@@ -31,6 +86,18 @@ int main(int argc, char *argv[]) {
+        std::chrono::duration_cast<std::chrono::microseconds>(elapsed).count() / 1000000.0;
+
+    renderer::Scene sceneToRender({Transform(cubeMesh, elapsedInSeconds)}, camera);
+    rendererObject.Render(sceneToRender, sdlRenderer);
```
Плохой дизайн. У тебя метод принимает данные, но не возвращает ничего. Это значит, что хуй пойми как у тебя данные и куда текут. Вот как понять направление данных в программе? Как понять кто именно занимается отрисовкой? Как понять, что в точности делает метод Render? Его название либо врет и он не только строит картинку, но и отображает ее, либо хуй пойми у тебя какие неявные зависимости. Не удивлюсь, если ты еще и аргументы по неконстантной ссылке принимаешь. Убрать это говно.

DimaTrushin commented on src/main.cpp:
```diff
@@ -31,6 +86,18 @@ int main(int argc, char *argv[]) {
+    auto elapsedInSeconds =
+        std::chrono::duration_cast<std::chrono::microseconds>(elapsed).count() / 1000000.0;
+
+    renderer::Scene sceneToRender({Transform(cubeMesh, elapsedInSeconds)}, camera);
```
Какого хуя у тебя время в секундах передается куда-то?

DimaTrushin replied:
Почему камера не мувается внутрь сцены? Почему у камеры не указываются координаты, при запихивании ее в сцену? Просто камера либо вообще ничего не знает про свою систему координат, либо у нее есть локальная система координат. И тебе надо уметь вложить камеру в сцену, это значит, что тебе надо указать стартовую точку и ортогональную матрицу для вложения локальной системы координат.

DimaTrushin replied:
Что вообще означает первыйы аргумент?

DimaTrushin commented on src/main.cpp:
```diff
@@ -1,26 +1,81 @@
 
-  SDL_SetRenderDrawColor(renderer.renderer, 0, 0, 0, 0);
-  SDL_RenderClear(renderer.renderer);
+  renderer::Renderer rendererObject(WINDOW_WIDTH, WINDOW_HEIGHT, std::move(color_pixel));
```
Renderer не должен принимать никакой call back для отрисовки где-то. Это НЕ его блять задача. Его задача лишь строить изображение. А куда ты его потом будешь выводить, это уже не его собачье дело.

Не бинди в конструкторе на объект случайные действия.

DimaTrushin commented on src/main.cpp:
```diff
@@ -1,26 +1,81 @@
-  // TODO: Rendering here :)
-  LOG(INFO) << "Finished drawing";
+  // for now camera is fixed to (0, 0, 0) looking at (0, 0, -1)
+  renderer::Camera camera(std::numbers::pi / 2, 0.1, 100.0);
```
Плохой интерфейс для конструктора камеры. Смысл всех этих волшебных констант совершенно не ясен.

DimaTrushin commented on src/main.cpp:
```diff
@@ -1,26 +1,81 @@
-  LOG(INFO) << "Finished drawing";
+  // for now camera is fixed to (0, 0, 0) looking at (0, 0, -1)
+  renderer::Camera camera(std::numbers::pi / 2, 0.1, 100.0);
+  renderer::Mesh cubeMesh = LoadCube();
```
Вот этим должен заниматься конструктор Application (или Model, если ты используешь MVC), потому что он ответственнен за конструирование и наполнение всех своих компонент.

DimaTrushin commented on src/main.cpp:
```diff
@@ -31,6 +86,18 @@ int main(int argc, char *argv[]) {
       }
     }
+
+    SDL_SetRenderDrawColor(sdlRenderer.renderer, 0, 0, 0, 0);
```
В библиотечных вызовах есть куча странных параметров, которые надо подставлять по умолчанию. В твоем коде не должно быть ненужных лишних параметров. Потому такие вызовы надо заворачивать в свои функции с нормальными названиями. Но это все равно должны быть объекты являющиеся частями view объектов.

DimaTrushin commented on src/scene/camera.hpp:
```diff
@@ -0,0 +1,16 @@
+
+struct Camera {
+  Camera(double fov, double z_near, double z_far) : fov(fov), z_near(z_near), z_far(z_far) {
+  }
```
У структуры НЕ должно быть конструкторов. Ты можешь воспользоваться aggregate initialization вида
```cpp
Camera camera1{.fov = 1, .z_near = 2, .z_far = 3};
```
И все.

DimaTrushin commented on src/scene/camera.hpp:
```diff
@@ -0,0 +1,16 @@
+namespace renderer {
+
+struct Camera {
+  Camera(double fov, double z_near, double z_far) : fov(fov), z_near(z_near), z_far(z_far) {
```
Плохой дизайн конструктора в целом, потому что у тебя ничего не мешает перепутать местами аргументы. И ничего не говорит тебе о смысле аргументов при вызове конструктора. Вот как понимать такой вызов
```cpp
Camera camera(1,2,3);
```
???

DimaTrushin commented on src/scene/camera.hpp:
```diff
@@ -0,0 +1,16 @@
+#pragma once
+
+#include <Eigen/Core>
```
Лишний include. Убрать.

DimaTrushin commented on src/scene/from_obj.cpp:
```diff
@@ -0,0 +1,73 @@
+
+  std::string file_line;
+
+  std::vector<Eigen::Vector3f> vert;
```
1) Не используй голые библиотечные типы. Если тебе придется поменять где-то тип вектора, то тебе придется это делать руками во всем коде. Лучше заведи прокси хедер а-ля Linalg.h, куда подключаешь Eigen, а в нем в своем namespace renderer объявляешь все нужные типы
```cpp
#include <Eigen/Core>

namespace renderer {
using Vector3f = Eigen::Vector3f;
...
}
```
И уже потом всюду используй свой прокси хедер.
2) У тебя тут Vector из float-ов используется. В камере у тебя double. Очень плохая тема смешивать разные типы в коде. Приведи все к единому виду.

DimaTrushin commented on src/scene/from_obj.cpp:
```diff
@@ -0,0 +1,73 @@
+  std::vector<Eigen::Vector3f> vert;
+  Mesh result{};
+
+  auto ReadOneCoordOfFace = [](std::stringstream& line) -> size_t {
```
Вот нахуя это тут надо писать? Что, блять, мешает сделать вспомогательную функцию внутри anonymous namespace? Ты просто забиваешь код более высокоуровневой функции кодом вызываемой функции. Ты лишаешься всех бонусов. Ты все равно вспомогательный код написал прямо в тексте высокоуровневой функции, потому не избавился от спагетти кода. И код у тебя перестал быть линейным как в спагетти коде и тебе все равно надо будет рыскать между вспомогательными функциями, тела которых к тому же еще и размазаны по самой функции. Убрать.

DimaTrushin commented on src/scene/mesh.hpp:
```diff
@@ -0,0 +1,18 @@
+
+struct Triangle {
+  using Vector = Eigen::Vector3f;
+  Vector pts[3]{};
```
Не надо использовать Си массивы. Что мешает использовать std::array?

DimaTrushin commented on src/scene/from_obj.cpp:
```diff
@@ -0,0 +1,73 @@
+        LOG(ERROR) << "triangle (face) is pointing at a non-existent vertex";
+        return std::nullopt;
+      }
+      result.triangles.emplace_back(Triangle{
```
1) Во-первых, если ты пишешь код вида
```cpp
obj.subobj.subsubobj.f();
```
Это ошибка дизайна. Это означает, что ты не предоставил интерфейс достаточный для работы с объектом obj и вместо этого копаешься в его кишках.
2) Во-вторых, метод emplace и его разновидности используются для того, чтобы напрямую передавать аргументы конструктору. Так как у тебя c++20, то теперь это работает и для aggregate initialization, то есть можно написать
```cpp
triangles.emplace_back({vert[v1], vert[v2], vert[v3]});
```
И еще я не понимаю, как это компилируется, когда у тебя "," стоит после }

DimaTrushin replied:
Что мешало выделить этот код в функцию читающую все треугольники, а потом мувнуть содержимое внутрь mesh?

DimaTrushin commented on src/scene/from_obj.cpp:
```diff
@@ -0,0 +1,73 @@
+  std::string file_line;
+
+  std::vector<Eigen::Vector3f> vert;
+  Mesh result{};
```
У тебя эта переменная начинает использоваться только в строчке 65. Зачем такое объявление заранее? Чтобы было сложнее читать код?

DimaTrushin commented on src/scene/from_obj.hpp:
```diff
@@ -0,0 +1,12 @@
+#pragma once
+
+#include <fstream>
+#include <glog/logging.h>
```
Эти хедеры тут лишние. Ты нигде не пользуешься в хедере этими хедерами. Тебе нужна тут только
```cpp
#include <string>
```
Эти надо перенести в cpp файл.

DimaTrushin commented on src/scene/from_obj.hpp:
```diff
@@ -0,0 +1,12 @@
+
+namespace renderer {
+
+std::optional<Mesh> LoadMeshFromObjFile(const std::string& filepath);
```
Плохая идея использвать голые std::string для названия файлов и путей. Лучше уж использовать std::filesystem::path

DimaTrushin commented on src/scene/from_obj.hpp:
```diff
@@ -0,0 +1,12 @@
+
+std::optional<Mesh> LoadMeshFromObjFile(const std::string& filepath);
+
+}
```
Кирпичи

DimaTrushin commented on src/scene/from_obj.cpp:
```diff
@@ -0,0 +1,73 @@
+      size_t v1, v2, v3;
+      std::tie(v1, v2, v3) = ReadOneFace(stream);
+      if (v1 >= vert.size() || v2 >= vert.size() || v3 >= vert.size()) {
+        LOG(ERROR) << "triangle (face) is pointing at a non-existent vertex";
```
Это уже 4-ый вложенный scope. Даже в ядре Linux запрещено использовать больше 3. Переделать.

DimaTrushin commented on src/scene/from_obj.cpp:
```diff
@@ -0,0 +1,73 @@
+      continue;
+    }
+
+    if (cmd == "v") {  // vertex
```
Комментарии в коде -- это всегда ошибка. Причем самое классное в этом комментарии то, что по его содержанию понятно в чем ошибка и как ее исправлять. Тут ты проверяешь какое-то непонятное условие внутри if и тебе приходится объяснять, что ты делаешь. А можно было бы создать вспомогательную функцию с нормальным названием и дернуть ее тут.
```cpp
bool is_next_vertex(const std::string& cmd) {
  /*implementation*/
}
if(is_next_vertex(cmd)) {
  ...
}
```
И у тебя уходит необходимость писать подобные комментарии, так как теперь появилась функци с нормальным названием.

DimaTrushin commented on src/scene/from_obj.cpp:
```diff
@@ -0,0 +1,73 @@
+          .pts = {vert[v1], vert[v2], vert[v3]},
+      });
+    } else {
+      LOG(WARNING) << "obj entry \"" << cmd << "\" not supported yet";
```
Тут лучше падать по assert-у, а не просто логать. Ибо раз ты вошел в ветку, в которую программа не должна была входить, ты где-то как программист накосячил и тебе все равно надо исправлять эту ошибку.

DimaTrushin commented on src/scene/from_obj.cpp:
```diff
@@ -0,0 +1,73 @@
+    } else if (cmd == "f") {
+      size_t v1, v2, v3;
+      std::tie(v1, v2, v3) = ReadOneFace(stream);
+      if (v1 >= vert.size() || v2 >= vert.size() || v3 >= vert.size()) {
```
Тут можно было бы тоже сделать функцию is_face_correct или что-то в этом духе.

DimaTrushin commented on src/renderer.hpp:
```diff
@@ -0,0 +1,67 @@
+
+namespace renderer {
+
+template <typename ColorPixel>
+  requires(std::invocable<ColorPixel, size_t, size_t, uint8_t, uint8_t, uint8_t>)
```
1) Renderer -- это компонент приложения. Любой компонент приложения должен быть конкретным классом, а не шаблоном. Ты можешь использовать шаблонный код в виде библиотечного кода для имплементации чего-то, но компоненты приложения всегда конкретные классы.
2) Кроме того, убрать нахуй этот сраный call back из рендерера. Его задача только в том, чтобы строить картинку. Все.

DimaTrushin commented on src/renderer.hpp:
```diff
@@ -0,0 +1,67 @@
+  requires(std::invocable<ColorPixel, size_t, size_t, uint8_t, uint8_t, uint8_t>)
+class Renderer {
+ public:
+  Renderer(size_t screen_width, size_t screen_height, ColorPixel&& color_pixel)
```
Ты используешь два одинаковых сырых типа в качестве подряд идущих аргументов функции. Тут есть проблема. что на стороне пользователя не ясно, что значат эти параметры. Например
```cpp
Renderer r(100, 200);
```
Вот что из этого высота, а что ширина? Кроме того, если ты их перепутаешь местами, нет никакого способа это обнаружить и придется долго вылавливать ошибки. Обычно используют strongly typed aliases, которые для целочисленных встроенных типов пишутся с помощью enum-ов так:
```cpp
enum Width : uint32_t;
enum Height : uint32_t;
class Renderer  {
public:
  Renderer(Width width, Height height);
  ...
};
```
И тогда на стороне пользователя будет
```cpp
Renderer  r(Width{100}, Height{200});
```
И мало того, что это теперь читается лучше, так еще если ты перепутаешь порядок аргументов, то на тебя ругнется компилятор. Что очень хорошо.

DimaTrushin replied:
Убрать color_pixel нахуй

DimaTrushin commented on src/renderer.hpp:
```diff
@@ -0,0 +1,67 @@
+               0, 0, z_scale, z_offset,
+               0, 0, -1, 0;
+
+    SDL_SetRenderDrawColor(sdlRenderer.renderer, 255, 255, 255, 255);
```
Renderer -- компонент ядра приложения. Ядро приложения никогда НЕ должно зависеть от GUI библиотек или других библиотек не относящихся к тому, что ядро делает. Убрать.

DimaTrushin replied:
Магические константы.

DimaTrushin commented on src/renderer.hpp:
```diff
@@ -0,0 +1,67 @@
+  }
+
+ private:
+  size_t screen_width_, screen_height_;
```
Никогда НЕ объявляй переменные в одной строке.

DimaTrushin commented on src/renderer.hpp:
```diff
@@ -0,0 +1,67 @@
+
+ private:
+  size_t screen_width_, screen_height_;
+  const ColorPixel color_pixel_;
```
НИКОГДА НЕ делай const поля в классе.

DimaTrushin commented on src/renderer.hpp:
```diff
@@ -0,0 +1,67 @@
+        color_pixel_(std::move(color_pixel)) {
+  }
+
+  void Render(const Scene& scene, /* WILL REMOVE: */ SDL::Renderer& sdlRenderer) {
```
Какого хуя метод Render не возвращает ничего? Я бы ожидал, что этот метод на вход получает:
1) сцену (включая все объекты, источники освещения и камеры)
2) идентификатор камеры, с которой надо рисовать
И возвращает класс содержащий картинку для отрисовки.

DimaTrushin commented on src/renderer.hpp:
```diff
@@ -0,0 +1,67 @@
+  void Render(const Scene& scene, /* WILL REMOVE: */ SDL::Renderer& sdlRenderer) {
+    // TODO: clipping
+    double e = 1 / tan(scene.camera_.fov / 2);
+    double aspect = static_cast<double>(screen_height_) / screen_width_;
```
Когда ты используешь static_cast в бизнес логике это означает, что ты передаешь типы данных не в том формате, в котором должен с ними работать. Это значит, что ты не написал нужные функции конвертации или не продумал интерфейсы объектов для взаимодействия и потому делаешь низкоуровневую ненужную работу руками каждый раз с нуля.

Судя по всему, тебе нужны функции конвертации из непрерывной системы координат в дискретную и обратно. Но я не уверен. Так же не ясно, почему за aspect_ratio не отвечает метод камеры? Ибо только камера знает, какие у нее соотношения сторон.

DimaTrushin commented on src/renderer.hpp:
```diff
@@ -0,0 +1,67 @@
+
+  void Render(const Scene& scene, /* WILL REMOVE: */ SDL::Renderer& sdlRenderer) {
+    // TODO: clipping
+    double e = 1 / tan(scene.camera_.fov / 2);
```
Почему это не метод камеры? Зачем тебе тогда камера, если ты все делаешь руками? Задача камеры как раз инкапсюлировать внутри своих методов всю работу по преобразованию координат.

DimaTrushin replied:
охуенное название переменной

DimaTrushin commented on src/renderer.hpp:
```diff
@@ -0,0 +1,67 @@
+    projMat << x_scale, 0, x_translation, 0,
+               0, y_scale, y_translation, 0,
+               0, 0, z_scale, z_offset,
+               0, 0, -1, 0;
```
1) Заведи using-ги для названия типов.
2) Это должна быть одна строчка
```cpp
Matrix4 projMat = makeProjMatrix(...);
```
Так же не ясно проекция на что это. Если это проекция на экран камеры, то почему камера не возвращает эту матрицу в своем методе?

DimaTrushin commented on src/renderer.hpp:
```diff
@@ -0,0 +1,67 @@
+    for (auto& mesh : scene.meshes_) {
+      auto projectedMesh = mesh.Transform(projMat);
+      for (auto& tri : projectedMesh.triangles) {
+        float x0 = (tri.pts[0][0] + 1) / 2 * screen_width_,
```
У тебя в камере лежат double, тут ты работаешь с float. Внимание вопрос: какого хуя?

DimaTrushin commented on src/renderer.hpp:
```diff
@@ -0,0 +1,67 @@
+    double e = 1 / tan(scene.camera_.fov / 2);
+    double aspect = static_cast<double>(screen_height_) / screen_width_;
+
+    double x_scale = aspect * e;
```
У тебя тут double, а в строчке 46 внезапно летят float. Какого хуя?

DimaTrushin commented on src/renderer.hpp:
```diff
@@ -0,0 +1,67 @@
+
+    double z_range_inverse = 1 / (scene.camera_.z_far - scene.camera_.z_near);
+    double z_scale = -(scene.camera_.z_far + scene.camera_.z_near) * z_range_inverse;
+    double z_offset = -2 * scene.camera_.z_near * scene.camera_.z_far * z_range_inverse;
```
Это все должны быть методы камеры

DimaTrushin commented on src/scene/mesh.hpp:
```diff
@@ -0,0 +1,18 @@
+struct Mesh {
+  std::vector<Triangle> triangles;
+
+  Mesh Transform(const Eigen::Matrix4d& mat) const;
```
1) Либо закрывай данные структуры в класс, либо делай глобальный метод, либо на крайний случай статический метод.
2) Ты не консистентный. У тебя в классах данные лежат в конце, в структурах в начале. Перейти к единому стандарту.

DimaTrushin commented on src/scene/scene.hpp:
```diff
@@ -0,0 +1,16 @@
+
+struct Scene {
+  Scene(const std::vector<Mesh>& meshes, const Camera& camera) : meshes_(meshes), camera_(camera) {
+  }
```
Убрать это говно

DimaTrushin commented on src/scene/scene.hpp:
```diff
@@ -0,0 +1,16 @@
+  }
+
+  std::vector<Mesh> meshes_;
+  Camera camera_;
```
По хорошему тебе надо связать с камерой вложенной в сцену ее систему координат, начальную точку и ортогональную матрицу для указания положения.

DimaTrushin commented on src/scene/scene.hpp:
```diff
@@ -0,0 +1,16 @@
+namespace renderer {
+
+struct Scene {
+  Scene(const std::vector<Mesh>& meshes, const Camera& camera) : meshes_(meshes), camera_(camera) {
```
Блять. Я же говорил, что НЕ надо принимать детали имплементации по константной ссылке. Принимай их по rvalue ссылке. А лучше сделай методы добавления меша и камеры и скрой это все в классе, а не структуре.

DimaTrushin commented on src/scene/mesh.cpp:
```diff
@@ -0,0 +1,25 @@
+//   ... but for now we're only loading a single cube so it's fine
+renderer::Mesh renderer::Mesh::Transform(const Eigen::Matrix4d& mat) const {
+  Mesh result;
+  result.triangles.reserve(triangles.size());
```
Создай вектор треугольников тут, а потом в конструктор result мувай вектор считанных треугольников
```cpp
std::vector<Triangle> triangles transform(mesh.triangles(), matrix);
Mesh result(std::move(triangles));
return result;
```
Все. И не надо писать в текущий mesh, просто возвращай этот mesh из функции и потом мувай данные туда, куда надо. Если же ты хочешь делать inplace, то можно поглотить mesh, изменить внутри и потом мувнуть наружу. Но ты этого тоже не делаешь.

DimaTrushin replied:
Как только ты у структур делаешь методы, то сразу невозможно отличить локальные переменные от переменных структуры, потому что у них нет _ в конце. Я тут чуть крышей не поехал, пока понял, что происходит.

DimaTrushin commented on src/scene/mesh.cpp:
```diff
@@ -0,0 +1,25 @@
+
+// THIS IS INEFFICIENT
+//   ... but for now we're only loading a single cube so it's fine
+renderer::Mesh renderer::Mesh::Transform(const Eigen::Matrix4d& mat) const {
```
Открой по нормальному namespace и пиши внутри него. Это сократит количество шума в коде. Сравни
```cpp
namespace renderer {
Mesh Mesh::Transform(const Eigen::Matrix4d& mat) const {
  ...
}
}
```

DimaTrushin commented on src/scene/mesh.cpp:
```diff
@@ -0,0 +1,25 @@
+      Eigen::Vector4d vec(orig[0], orig[1], orig[2], 1);
+      Eigen::Vector4d vecTransformed = mat * vec;
+      double w = vecTransformed[3];
+      if (std::abs(w) > EPS) {
```
1) Опять 4 вложенных scope.
2) Заведи функцию на сравнение с нулем типа almost_equal(a, b) и almost_zero(a). И пусть EPS будет внутренним параметром этой функции (или функцтора).

DimaTrushin commented on src/scene/mesh.cpp:
```diff
@@ -0,0 +1,25 @@
+        vecTransformed *= 1 / w;
+      }
+      renderer::Triangle::Vector newVec(vecTransformed[0], vecTransformed[1], vecTransformed[2]);
+      result.triangles.back().pts[i] = newVec;
```
Блять

DimaTrushin commented on src/scene/mesh.cpp:
```diff
@@ -0,0 +1,25 @@
+renderer::Mesh renderer::Mesh::Transform(const Eigen::Matrix4d& mat) const {
+  Mesh result;
+  result.triangles.reserve(triangles.size());
+  for (auto& tri : triangles) {
```
почему не константная ссылка?

DimaTrushin commented on src/scene/mesh.cpp:
```diff
@@ -0,0 +1,25 @@
+#include <scene/mesh.hpp>
+
+constexpr double EPS = 1e-10;
```
Никогда не пиши в global namespace, НИКОГДА!

DimaTrushin commented on src/scene/mesh.cpp:
```diff
@@ -0,0 +1,25 @@
+  result.triangles.reserve(triangles.size());
+  for (auto& tri : triangles) {
+    result.triangles.emplace_back();
+    for (size_t i = 0; i < 3; ++i) {
```
Почему не выделить это в отдельную функцию, куда бы ты передал triangles.back()?

DimaTrushin reviewed (changes requested):
Обрати внимание, что я не повторял замечания. Если одна и та же ошибка всречалась несколько раз, я комментировал лишь единожды. Потому я ожидаю, что ты поправишь все аналогичные замечания всюду.

Commit:
rewritten wireframe

Commit:
added imgui

Commit:
Abstracted away scene, changed projection matrix

Also added axes x and y (as meshes) for reference.
Also fixed parsing obj files.
Also deleted imgui.ini.

Commit:
implemented clipping

Commit:
added camera movement, added controller, added teapot object to scene

Commit:
small change to datalist parser: added comments

Commit:
added culling to non-wireframe mode

Commit:
now filling triangles in non-wireframe mode. no z-buffer yet.

Commit:
z-buffer added, weird artifacts on screen

Commit:
1/2 artefacts fixed. z-buffer still acting wonky

Commit:
quick fix: added static_cast for double to float conversion

Commit:
quick fix: change lshift -> z for going down

Commit:
fixed incorrect depth calculating issue

Commit:
small tweaks

Commit:
added parallelism to a small part of code

Went from 70 fps on Utah teapot to 144 fps
on my machine all while consuming about 10-15%
cpu. Tried OpenMP, got worse results with 100%
cpu usage.

Commit:
colored lighting

Commit:
added texture(s)

Commit:
light source is not rotating

Commit:
removed OpenMP

Commit:
on new scene no need to modify app.cpp

Commit:
moved all of image & texture to hpp

with code in cpp file, in separate translation
unit, operator[] wasn't inlined (I checked asm output).
therefore this made it possible to inline the hot
path (FillScanline)
result fps went ~80 -> ~110

Commit:
fixed textures

Commit:
little cleaning up before report deadline

Commit:
added a cottage model

SecondSon1 committed:
cleaning code up

DimaTrushin commented on src/main.cpp:
```diff
@@ -1,37 +1,46 @@
+  InitializeGlobalObjects(argv[0]);
 
-  SDL_RenderPresent(renderer.renderer);
+  renderer::MoveToRoot();
```
Какого хрена это тут? И что эта хрень вообще делает?

SecondSon1 replied:
MoveToRoot() двигает CWD (current working directory) в корень проекта. Так как хрен пойми откуда юзер или ide запустит приложуху, а искать data как-то надо.

DimaTrushin commented on src/main.cpp:
```diff
@@ -1,37 +1,46 @@
 
-  SDL::Window window(WINDOW_TITLE, WINDOW_WIDTH, WINDOW_HEIGHT, 0);
-  SDL::Renderer renderer = window.GetRenderer(nullptr);
+renderer::ScenePtr CreateExampleScene(std::vector<renderer::Mesh> &&mesh, renderer::Texture &&tex) {
```
Какого хрена эта деталь имплементации у тебя снаружи висит? Почему ты не мог имплементировать сцену так, чтобы мув для нее был дешовой операций, чтобы не надо было заворачивать все в unique_ptr?

DimaTrushin commented on src/view.hpp:
```diff
@@ -0,0 +1,38 @@
+ private:
+  SDL::Window window_;
+  SDL::Renderer renderer_;
+  bool options_drawn_;
```
Не использовать bool или int и прочие флаги со встроенными типами. Лучше делать enum class с хорошими названиями состояний, которые еще дополнительно не может конвертнуться во что-то лишнее.

DimaTrushin commented on src/app.hpp:
```diff
@@ -0,0 +1,53 @@
+  util::Timer timer_;
+  util::TimingsMeasurer frame_timings_measurer_;
+
+  bool is_running_ = false;
+  bool is_paused_ = true;
+  bool is_wireframe_ = false;
```
Лучше сделать enum со статусом и хранить в одной маске все флаги. Кроме того, от этого флага у тебя скорее всего меняется обработчик события. Потому можно вместо флага хранить сам обработчик.

SecondSon1 replied:
Здесь эти bool передаются напрямую в ImGui. В данной библиотеке если у тебя есть чекбокс, ты не можешь поставить на него callback. Ты просто передаешь ему указатель на bool и он его поменяет когда пользователь на него кликнет. По крайней мере is_paused_ и is_wireframe_ необходимы, так как они напрямую соответствуют чекбоксам в менюшке в приложении.

DimaTrushin commented on src/app.hpp:
```diff
@@ -0,0 +1,53 @@
+
+ private:
+  Renderer renderer_;
+  std::unique_ptr<Scene> scene_;
```
Нафига тут unique_ptr? Выглядит как выворачивание кишков у сцены. Если это для дешового мува, то почему нельзя было это запрятать внутрь имплементации сцены7

SecondSon1 replied:
Scene это абстрактный класс, который определяет пользователь. В этом задумка: пользователь как-то обрабатывает mesh'ы во время инициализации сцены в конструкторе, а дальше на каждый tick сцены может как-то поменять сцену. Идея в том, что пользователь может написать несколько сцен и как-то их менять, правда механизма сделать это в runtime я не сделал.

DimaTrushin commented on src/app.hpp:
```diff
@@ -0,0 +1,53 @@
+
+namespace renderer {
+
+using ScenePtr = std::unique_ptr<Scene>;
+using SceneCreator = std::function<ScenePtr(std::vector<Mesh>&&, Texture&&)>;
```
Не думаю, что эти вещи должны быть тут в публичном доступе. Это детали имплементации класса, которые надо запрятать внутри.

SecondSon1 replied:
Мне кажется, используя их проще писать функцию, которые позже передается в конструктор Application.

DimaTrushin commented on src/app.hpp:
```diff
@@ -0,0 +1,53 @@
+  std::unique_ptr<Scene> scene_;
+  ImageWithDepth last_image_;
+  Camera camera_;
+  Controller ctl_;
+  View view_;
```
Нарушение уровня абстракции. У тебя есть контроллер и view. Но при этом нет модели как таковой. У тебя по сути рендерер, сцена, и камера -- это детали имплементации единой модели. А last_image -- это вообще должны быть данные которыми модель оповещает view. Ты распилил модель и выпустил ее кишки наружу.

SecondSon1 replied:
View это действительно View из MVC. Но Controller это скорее абстракция над методом управления, а не концепт из MVC. Этот класс отвечает за получение событий, поступивших с контроллера (клавомышь, джойстик, геймпад и т д) и как-то их обработать.

DimaTrushin commented on src/app.hpp:
```diff
@@ -0,0 +1,53 @@
+  Camera camera_;
+  Controller ctl_;
+  View view_;
+  util::Timer timer_;
+  util::TimingsMeasurer frame_timings_measurer_;
```
Зачем тут таймер я еще могу догадаться. А вот зачем тут вторая переменная я пока не понимаю. Это видимо величиная измеряющая временной интервал. Ее точно надо запоминать между вызовами?

SecondSon1 replied:
TimingsMeasurer хранит не последний тайминг, а сразу много штук. Таким образом он не выдает последнее известное значение, а вместо этого он берет среднее по некоторому периоду времени, чтобы fps не скакал и в целом был более информативным.

DimaTrushin commented on src/app.hpp:
```diff
@@ -0,0 +1,53 @@
+  std::unique_ptr<Scene> scene_;
+  ImageWithDepth last_image_;
+  Camera camera_;
+  Controller ctl_;
+  View view_;
```
Так же меня смущают механизмы связи объектов. Я еще их не видел и меня это пугает.

SecondSon1 replied:
Application это тупо мастер-объект, Controller это печальная коллизия имен, подробнее описал выше.

DimaTrushin commented on src/app.cpp:
```diff
@@ -0,0 +1,137 @@
+      camera_(InitializeCamera()),
+      ctl_(camera_),
+      view_(kWindowWidth, kWindowHeight, std::string(kWindowTitle)) {
+}
```
У тебя нет автоматического соединения view и модели. Ты просто делаешь application мастер объектом, который вручную передает данные. Потому не было необходимости вообще городить огород из view и controller-а.

DimaTrushin commented on src/app.cpp:
```diff
@@ -0,0 +1,137 @@
+  constexpr size_t kTargetFPS = 200;
+  constexpr util::Timer::SecondsUnit kSecondsPerFrame = 1.0 / kTargetFPS;
+
+  is_running_ = true;
```
Судя по всему этот флаг это локальная переменная метода, а не application

SecondSon1 replied:
Здесь проблема в том, что я хочу выходить из приложения извне метода, в частности в HandleEvents. Делать целую event-driven архитектуру для одного сообщения - о выходе - показалось избыточным, и я решил сэкономить и себе, и тебе время, заменив его одним флагом. А возвращать из HandleEvents какое-либо значение, сигнализирующее о том, что пора закругляться звучит как плохой план.

DimaTrushin commented on src/app.cpp:
```diff
@@ -0,0 +1,137 @@
+    }
+
+    view_.BeginFrame();
+    auto seconds_per_last_frame = timer_.Lap();
+    frame_timings_measurer_.AddTiming(seconds_per_last_frame);
```
Что мешало сделать один объект timer, который будет иметь один единственный метод -- выдать время от предыдущего измерения. И все, у тебя просто одна строчка вызова в начале цикла. И не надо руками все эти вычисления делать.

SecondSon1 replied:
Объяснил выше - Timer замеряет один тайминг, а TimingsMeasurer хранит сразу много их и выдает среднее.

DimaTrushin commented on src/app.cpp:
```diff
@@ -0,0 +1,137 @@
+    ctl_.Advance(seconds_per_last_frame);
+
+    last_image_ = RenderImage();
+    DrawOptions();
```
Вот тут плохой метод. Во-первых, он ничего не принимает и ничего не возвращает. Во-вторых, не понятно куда он пишет во view или нет. В-третьих, нарушение уровня абстракции, так как код выше и ниже работает напрямую с полями приложения.

SecondSon1 replied:
Это так работает ImGui. Если что-то надо в нем нарисовать, то вызываем функцию ```ImGui::DrawText("text");``` Он не может ничего возвращать, да и не пишет он никуда кроме какого-то глобального состояния самого ImGui.

DimaTrushin commented on src/app.cpp:
```diff
@@ -0,0 +1,137 @@
+    view_.BeginFrame();
+    auto seconds_per_last_frame = timer_.Lap();
+    frame_timings_measurer_.AddTiming(seconds_per_last_frame);
+    if (!is_paused_) {
+      scene_->Advance(seconds_per_last_frame);
+    }
+    ctl_.Advance(seconds_per_last_frame);
```
Какого хрена тут происходит? Ты управляешь в двух местами движением сцены? Что за дичь?

SecondSon1 replied:
Advance у сцены - виртуальный метод, который перегружает пользователь. Он как-то двигает сцену, причем мы не знаем (да и не хотим знать -- это дело пользователя) что он сделает со сценой.
Advance у controller - (соглашусь, не лучший нейминг) метод, который оповещает контроллер о том, что пора как-то повлиять на сцену. На практике только двигает камеру.

DimaTrushin commented on src/app.cpp:
```diff
@@ -0,0 +1,137 @@
+    view_.BeginFrame();
+    auto seconds_per_last_frame = timer_.Lap();
+    frame_timings_measurer_.AddTiming(seconds_per_last_frame);
+    if (!is_paused_) {
+      scene_->Advance(seconds_per_last_frame);
+    }
```
По хорошему у тебя должен быть аниматор сцены, и он уже внутри себя знает на паузе анимация или нет. И тут ты лишь ему скармливаешь информацию о прошедшем времени, чтобы он сам принял решение, что ему делать.

DimaTrushin commented on src/app.cpp:
```diff
@@ -0,0 +1,137 @@
+}  // namespace
+
+renderer::Application::Application(SceneCreator&& scene_creator)
+    : renderer_(kWindowWidth, kWindowHeight),
+      scene_(LoadScene(std::move(scene_creator))),
+      last_image_(kWindowWidth, kWindowHeight),
+      camera_(InitializeCamera()),
+      ctl_(camera_),
+      view_(kWindowWidth, kWindowHeight, std::string(kWindowTitle)) {
+}
```
Мне интересно, если я укажу для всех трех объектов renderer, last_image и view разные размеры, оно упадет или сработает? А если сработает, то корректно или нет? Просто я не люблю синхронизацию данных в коде, она делает его хрупким и я хочу понять есть ли она у тебя.

SecondSon1 replied:
Я бы предпочел, чтобы оно свалилось, но не уверен, что это будет так. А вообще так задумано. Не обязательно чтобы renderer имел тот же размер, что и view. view отвечает за окно и его ширина и высота - соответственно ширина и высота окна. У рендерера это ширина и высота изображения результата. Ну а last_image тут просто заполняется черным, можно в теории было вообще его не инициализировать, так как это конкретное изображение не увидит никто, и на практике у него всегда будут такие же высота и ширина, с которыми был инициализирован renderer, который это изображение и вернул.

DimaTrushin commented on src/app.cpp:
```diff
@@ -0,0 +1,137 @@
+
+}
+
+std::unique_ptr<Scene> Application::LoadScene(SceneCreator&& scene_creator) const {
```
А нахрена вообще этот scene_creator передавать в Application снаружи? Почему нельзя тут его создать внутри функции LoadScene?
И почему его надо мувать? Он что тяжелый? Это странно, ибо функторы должны быть легкими.

SecondSon1 replied:
Конструкторы не бывают виртуальными( При этом хочется, чтобы подгружать файлы с диска все еще было задачей Application, а не пользовательского Scene. Вот и получаются два стула - либо передать функтор, который инициализирует объект сцены, либо делать метод Init. Я выбрал первый, но это было после полуночи. Сейчас бы я так не сделал.

DimaTrushin commented on src/app.cpp:
```diff
@@ -0,0 +1,137 @@
+  DataLoader loader(kDatalistFilePath);
+  std::vector<Mesh> objects = loader.GetAllObjects();
+
+  std::string path_to_tex = "data/tex_all.jpg";
+  std::filesystem::path fs_path(path_to_tex);
+  Texture texture = TextureReader::LoadImage(fs_path);
```
Почему это не одна строчка вида
```cpp
Texture texture = TextureReader::Load(k_path_to_tex);
```
??

DimaTrushin commented on src/app.cpp:
```diff
@@ -0,0 +1,137 @@
+  return scene_creator(std::move(objects), std::move(texture));
+}
+
+Camera Application::InitializeCamera() const {
```
Почему это не статическая функция?

DimaTrushin commented on src/app.cpp:
```diff
@@ -0,0 +1,137 @@
+
+Camera Application::InitializeCamera() const {
+  return {
+      .fov_degrees_ = 90.0,
```
Судя по всему это поля структуры. У структуры нет необходимости в _ в названии переменных.

SecondSon1 replied:
Я раньше не ставил _ в конце и ты, насколько я помню, придрался к этому, поэтому я стал их ставить.

DimaTrushin commented on src/app.cpp:
```diff
@@ -0,0 +1,137 @@
+  view_.DrawCameraOptions(camera_);
+}
+
+void Application::Quit() {
```
Вот этот метод врет. Он не выходит из runtime loop, он выставляет флаг о том, что надо выйти из этого loop-а. Это что-то вроде deffer_quit.

DimaTrushin commented on src/controller.hpp:
```diff
@@ -0,0 +1,51 @@
+  void MouseMoved(SDL::MouseMotionEvent mouse_motion_event);
+  void MouseWheelMoved(SDL::MouseWheelEvent mouse_wheel_event);
+
+  double* GetCameraSpeedPtr();
+  double* GetCameraRotationSpeedPtr();
```
блять

SecondSon1 replied:
Это тоже для ImGui, по аналогии с bool, если надо сделать scrollbar, то передаешь указатель на float/double и он сам втихаря его обновит.

DimaTrushin commented on src/controller.hpp:
```diff
@@ -0,0 +1,51 @@
+
+class Controller {
+ public:
+  Controller(Camera& camera);
```
Лучше принимай указатель. Это важно не на стороне контроллера, а на стороне того, кто использует контроллер. Будет ясно, что камера туда не как разовый параметр передается.

DimaTrushin commented on src/controller.hpp:
```diff
@@ -0,0 +1,51 @@
+ public:
+  Controller(Camera& camera);
+
+  void Advance(util::Timer::SecondsUnit elapsed);
```
Заведи юзинги в начале класса, которые бы убирали бы все эти лишние квалификаторы доступа через пространства имен. Это повысит читаемость кода.

DimaTrushin commented on src/controller.cpp:
```diff
@@ -0,0 +1,105 @@
+}  // namespace
+
+Controller::Controller(Camera& camera)
+    : camera_(&camera), held_(std::make_unique<std::array<bool, Impl::kAmountOfScancodes>>()) {
+  std::fill(held_->begin(), held_->end(), false);
```
Вот тут видно, что ты "носишь недостаточное количество шляп". У тебя не достаточное количество вспомогательных классов. Ты выпотрошил наружу кишки объекта, который ты не написал, и который должен зниматься обработкой клавиш. Ты тут делаешь нерелевантную мелкую работу, которая должна быть завернута внутрь специального класса, а точнее его конструктора.

DimaTrushin commented on src/controller.cpp:
```diff
@@ -0,0 +1,105 @@
+void Controller::MouseWheelMoved(SDL::MouseWheelEvent mouse_wheel_event) {
+}
+
+double* Controller::GetCameraSpeedPtr() {
+  return &camera_speed_;
+}
+
+double* Controller::GetCameraRotationSpeedPtr() {
+  return &camera_rotation_speed_;
+}
```
У меня кроме вопроса "нахуя, а главное зачем?" ничего не возникает.

SecondSon1 replied:
Объяснил выше, в комментарии к объявлениям функций

DimaTrushin commented on src/controller.cpp:
```diff
@@ -0,0 +1,105 @@
+  auto rotation_transform = Eigen::AngleAxisd(roll, kNegZAxis) * Eigen::AngleAxisd(pitch, kXAxis) *
+                            Eigen::AngleAxisd(yaw, kYAxis);
+
+  Quaternion rot_quat(rotation_transform);
```
блять

SecondSon1 replied:
Я не знаю, как работать с кватернионами. Сделал как знал, зато работает. Плюс вызывается только раз в кадр, так что на скорость работы не влияет.

DimaTrushin commented on src/controller.cpp:
```diff
@@ -0,0 +1,105 @@
+
+int Controller::ScancodesSumTotal(std::initializer_list<SDL::Scancode> positive,
+                                  std::initializer_list<SDL::Scancode> negative) {
+  bool any_positive = false;
+  bool any_negative = false;
+  for (SDL::Scancode scancode : positive) {
+    any_positive |= (*held_)[scancode];
+  }
+
+  for (SDL::Scancode scancode : negative) {
+    any_negative |= (*held_)[scancode];
+  }
+
+  int result = any_positive - any_negative;
```
Во-первых, у тебя есть стандартные алгоритмы для вычисления any. Во-вторых, конвертация bool к int, спасибо.

DimaTrushin commented on src/controller.cpp:
```diff
@@ -0,0 +1,105 @@
+void Controller::AdvancePositionVertical(util::Timer::SecondsUnit elapsed) {
+  const Vector3d vert_offset = camera_->GetUpDirection() * camera_speed_;
+  int up = ScancodesSumTotal({SDL_SCANCODE_SPACE}, {SDL_SCANCODE_Z});
+  camera_->Offset(vert_offset * (up * elapsed));
```
- Плохо выбрано название метода.
- Плохо выбрано название переменных, у тебя vert_offset и up тут присутствуют. Складывается впечатление. что ты дважды указал направление вверх. И я не понимаю, что ты делаешь

SecondSon1 replied:
Название метода - глагол Offset. Он делает, что и должен, то есть сдвигает камеру на переданный вектор.

Метод делает следующее: он берет направление вверх из камеры, домножает его на скорость камеры. Дальше, ScancodesSumTotal возвращает 0 если либо не нажаты данные кнопки, либо нажаты обе. Возвращает 1 если нажата только первая и -1 если нажата только вторая. То есть, как метод говорит: каждая клавиша, можно представить, как будто вносит свой вклад. Пробел: +1, Z: -1. Если они нажаты обе, то они друг друга нивелируют и камера не двигается.

DimaTrushin commented on src/fs/root_finder.cpp:
```diff
@@ -0,0 +1,36 @@
+
+}  // namespace
+
+void MoveToRoot() noexcept {
```
Что это за хрень?

SecondSon1 replied:
Объяснил в самом верху, у коммента при использовании этой функции в main.

DimaTrushin commented on src/fs/obj_parser.hpp:
```diff
@@ -0,0 +1,45 @@
+  Mesh ConstructMesh();
+
+ private:
+  std::istream* stream_;
```
Какого хрена тут указатель на поток?

SecondSon1 replied:
Это parser. Он хранит поток и читает из него по ходу парсинга.

DimaTrushin commented on src/fs/texture_reader.cpp:
```diff
@@ -0,0 +1,64 @@
+      buf_(std::exchange(other.buf_, nullptr)) {
+}
+
+Texture::~Texture() {
+}
```
Это называется = default;

DimaTrushin commented on src/fs/texture_reader.hpp:
```diff
@@ -0,0 +1,63 @@
+  std::unique_ptr<impl::Texture> impl_;
+  Width width_;
+  Height height_;
+  unsigned char* buf_;
```
Какого хрена у тебя тут указатель на чистый буффер? А для чего у тебя impl_ тогда?

SecondSon1 replied:
Тут я юзаю pimpl, но тогда любой доступ в поле класса означал бы не только вызов функции без возможности inline, но и двойное разыменование: сначала impl_, потом buffer. Поэтому я дублирую сюда данные impl, чтобы исключить это двойное разыменование. Unique_ptr владеет и при вызове деструктора удалит данные, а buf_ просто содержит их копию чтобы работало быстрее.

DimaTrushin commented on src/fs/texture_reader.hpp:
```diff
@@ -0,0 +1,63 @@
+  Texture(Texture&& other) noexcept;
+  ~Texture();
+
+  Pixel operator[](Index ind) const {
```
Почему имплементация в хедере?

SecondSon1 replied:
Как и в классе ImageWithDepth, потому что operator[] хочется inline'ить, пришлось поместить сюда. Аналогичное действие в ImageWithDepth в свое время улучшило фпс с 70-80 до 100+ (я посмотрел асм и он не смог его inline'ить пока я не перенес имплементацию в хэдер).

DimaTrushin commented on src/fs/texture_reader.hpp:
```diff
@@ -0,0 +1,63 @@
+  Pixel operator[](Index ind) const {
+    assert(ind.row_ < GetHeight());
+    assert(ind.col_ < GetWidth());
+    size_t row_from_top = static_cast<size_t>(GetHeight()) - ind.row_ - 1;
```
Странно, что у тебя конвертация к подлежащему типу не происходит автоматом, ты не верно видимо реализовал эту парадигму.

SecondSon1 replied:
GetHeight() возвращает тип Height, ind.row_ имеет тип Row. Clang (на линуксе и маке) жаловался, что я вычитаю из enum'а enum, что нехорошо. Это был только warning, но я старался и их тоже убрать.

DimaTrushin commented on src/util/parallelism.hpp:
```diff
@@ -0,0 +1,27 @@
+  T* begin = &result[0];
+#if defined(__clang__) && defined(__apple_build_version__)
+  // Apple Clang
+  std::fill(begin, begin + sz, elem);
```
получается. что название метода врет, раз тут у тебя нет параллелизации. И почему эта радость тут возникает?

SecondSon1 replied:
Нет параллелизации только в случае Apple Clang, так как [как сказано здесь](https://en.cppreference.com/w/cpp/compiler_support/17) (строчка "Parallel algorithms and execution policies"), Apple Clang не поддерживает std::execution::par_unseq, который имеется в стандарте C++17. Поэтому здесь я, в этом случае, просто вызываю дефолтный fill как fallback.

DimaTrushin commented on src/util/parallelism.hpp:
```diff
@@ -0,0 +1,27 @@
+
+#include <memory>
+#include <execution>
+#include <glog/logging.h>
```
лишний хедер

DimaTrushin commented on src/util/drawing_primitives.hpp:
```diff
@@ -0,0 +1,44 @@
+                               Pixel lighting_color);
+
+ private:
+  ImageWithDepth* img_;
+  const Texture* tex_;
```
блять

SecondSon1 replied:
А здесь что не так?

DimaTrushin commented on src/scene/scene.hpp:
```diff
@@ -0,0 +1,28 @@
+class Scene {
+ public:
+  Scene(std::vector<Mesh>&& meshes, Lighting&& lighting, Texture&& tex);
+  virtual ~Scene() = default;
```
Блять

DimaTrushin reviewed (changes requested):
Добавил несколько замечаний. Ты мог бы написать сильно лучше.
